src/levanter/models/llama.py

```diff
@@ -471,18 +472,21 @@ def init(cls, Vocab: Axis, config: LlamaConfig, *, key) -> "LlamaLMHeadModel":
     def __call__(
         self,
         input_ids: NamedArray,
-        attn_mask: Optional[NamedArray] = None,
+        attn_mask: Optional[Union[NamedArray, AttentionMask]] = None,
         *,
         key=None,
     ) -> NamedArray:
         """
         Args:
             input_ids (NamedArray): [batch, position]
                 Indices of input sequence tokens in the vocabulary.
-            attn_mask (NamedArray, optional): [batch, position, seq_len]
+            attn_mask (Union[NamedArray, AttentionMask], optional): [batch, position]
                 Mask to avoid performing attention on the padding token indices of the encoder input.
+                The attn_mask from training pipeline may be an AttentionMask object instead of NamedArray
         """
         x = self.embeddings.embed(input_ids)
+        if isinstance(attn_mask, AttentionMask):
+            attn_mask = attn_mask.materialize()
```

Comment:
The attn_mask has to be a NamedArray; otherwise, it would cause an AssertionError in `hnn.attention.dot_product_attention`:

```
File "/home/ivan/levanter/src/levanter/models/llama.py", line 269, in __call__
    attn_output = hnn.attention.dot_product_attention(c.Pos, c.KeyPos, c.HeadSize, q, k, v, mask)
  File "/home/ivan/haliax/src/haliax/nn/attention.py", line 111, in dot_product_attention
    weights = dot_product_attention_weights(KeySize, KPos, query, key, mask, bias, attention_dtype, precision)
  File "/home/ivan/haliax/src/haliax/nn/attention.py", line 65, in dot_product_attention_weights
    weights = haliax.where(mask, weights, -1e9)
  File "/home/ivan/haliax/src/haliax/ops.py", line 41, in where
    assert isinstance(condition, NamedArray)
jax._src.traceback_util.UnfilteredStackTrace: AssertionError

The stack trace below excludes JAX-internal frames.
```

Reply:
probably should be a better error, but these "structured" masks are pretty important for ram usage when dealing with flash attention...

Reply:
@dlwh What would you recommend? Shall we pass the CausalMask from LMHead to LlamaAttention and materialize there? Or handle them in Haliax.attention?

Reply:
I have switched from CasualMask type to AttentionMask type. 
I also move the step of materialize inside LlamaAttention module, following the implementation at test_flash_attention [here](https://github.com/stanford-crfm/levanter/blob/f72a4c9018b90344dce56c25b39b69ee100f63a8/tests/test_flash_attention.py#L42).

Reply:
sounds good, thanks!